lib-extra/src/main/java/com/diffplug/spotless/extra/config/MavenCoordinates.java

```diff
@@ -0,0 +1,223 @@
+
+	/** Add missing coordinates and overwrite existing ones */
+	void update(Coordinate... coordinates) {
+		for (Coordinate coordinate : coordinates) {
+			this.coordinates.remove(coordinate); //Update existing
+			this.coordinates.add(coordinate);
+		}
+	}
+
+	/**
+	 * Add coordinates if missing.
+	 * Expecting reference to ASCII content.
+	 * Each line must correspond to the {@link Coordinate#FORMAT}.
+	 */
+	void add(URL coordinates) {
+		try {
+			BufferedReader reader = new BufferedReader(new InputStreamReader(coordinates.openStream(), "UTF-8"));
+			try {
+				List<Coordinate> additionalCoordinates = reader.lines().map(l -> new Coordinate(l)).collect(Collectors.toList());
+				this.coordinates.addAll(additionalCoordinates);
+			} finally {
+				reader.close();
+			}
+		} catch (IOException e) {
+			throw new UserArgumentException(coordinates, "Cannot access URL for optaining Maven coordinates.", e);
+		}
+	}
+
+	/** Add coordinates if missing */
+	void add(Coordinate... coordinates) {
+		for (Coordinate coordinate : coordinates) {
+			this.coordinates.add(coordinate);
+		}
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (obj instanceof MavenCoordinates) {
+			MavenCoordinates other = (MavenCoordinates) obj;
+			//Compare restrictions and not only the dependency
+			List<Coordinate> otherRestrictions = new ArrayList<Coordinate>(other.coordinates);
+			List<Coordinate> myRestrictions = new ArrayList<Coordinate>(coordinates);
+			return otherRestrictions.equals(myRestrictions);
+		}
+		return false;
+	}
+
+	@Override
+	public int hashCode() {
+		return coordinates.hashCode();
+	}
+
+	@Override
+	public String toString() {
+		return coordinates.stream().map(c -> c.toString()).collect(Collectors.joining(System.lineSeparator()));
+	}
+
+	static class Coordinate implements Serializable {
+		// Not used, only the serialization output is required to determine whether the object has changed
+		private static final long serialVersionUID = 1L;
+
+		// Format of a dependency version restriction (type is always JAR)
+		public static final String FORMAT = "<groupId>:<artifactId>[:packaging][:classifier]:<versionRestriction>";
+		private final String groupId;
+		private final String artifactId;
+		private final String packagingOrNull;
+		private final String classifierOrNull;
+		private final String versionRange;
+		private String dependency = null;
+
+		/**
+		 * Converts the format {@code <groupId>:<artifactId>[:packaging][:classifier]} dependency and
+		 * {@code <versionRestriction>} value
+		 */
+		Coordinate(String dependency, String version) {
+			this(String.format("%s:%s", dependency, version));
+		}
+
+		/** Creates version restriction from string. String format is {@value #FORMAT}. */
+		Coordinate(String coordinate) {
+			if (null == coordinate) {
+				throw new UserArgumentException(null, "Maven coordinate");
+			}
+			LinkedList<String> dependencyParts = new LinkedList<String>(
+					Arrays.asList(coordinate.split(":", -1)));
```

Comment:
Prefer `ArrayList` to `LinkedList`, unless you have a JMH benchmark which proves that `LinkedList` actually performs better.

Reply:
:-1: 
ArrayList does not support pollLast/pollFirst.

Reply:
Ah, in that case, try `Deque` and `ArrayDeque` instead. For queue- and stack-like operations like `pollFirst`/`pollLast`, ArrayDeque is nearly always the best implementation to use. To quote the ArrayDeque class javadoc:

> This class is likely to be faster than Stack when used as a stack, and faster than LinkedList when used as a queue.